xbmc/addons/interfaces/General.cpp

```diff
@@ -0,0 +1,202 @@
+}
+
+char* Interface_General::get_dvd_menu_language(void* kodiBase)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_General::%s - invalid data (addon='%p')", __FUNCTION__, addon);
+    return nullptr;
+  }
+
+  std::string string = g_langInfo.GetDVDMenuLanguage();
+  char* buffer = strdup(string.c_str());
+  return buffer;
+}
+
+void Interface_General::queue_notification(void* kodiBase, const int type, const char* message)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || message == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_General::%s - invalid data (addon='%p', message='%p')", __FUNCTION__, addon, message);
+    return;
+  }
+
+  switch (type)
+  {
+    case QUEUE_WARNING:
+      CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Warning, addon->Name(), message, 3000, true);
+      CLog::Log(LOGDEBUG, "Interface_General::%s - %s - Warning Message: '%s'", __FUNCTION__, addon->Name().c_str(), message);
+      break;
+  
+    case QUEUE_ERROR:
+      CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Error, addon->Name(), message, 3000, true);
+      CLog::Log(LOGDEBUG, "Interface_General::%s - %s - Error Message : '%s'", __FUNCTION__, addon->Name().c_str(), message);
+      break;
+  
+    case QUEUE_INFO:
+    default:
+      CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, addon->Name(), message, 3000, false);
+      CLog::Log(LOGDEBUG, "Interface_General::%s - %s - Info Message : '%s'", __FUNCTION__, addon->Name().c_str(), message);
+      break;
+  }
+}
+
+void Interface_General::queue_notification_from_type(void* kodiBase, const int type, const char* caption,
+                                                     const char* description, unsigned int displayTime,
+                                                     bool withSound, unsigned int messageTime)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || caption == nullptr || description == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_General::%s - invalid data (addon='%p', caption='%p', description='%p')", __FUNCTION__, addon, caption, description);
+    return;
+  }
+  
+  CGUIDialogKaiToast::eMessageType usedType;
+  switch (type)
+  {
+  case QUEUE_WARNING:
+    usedType = CGUIDialogKaiToast::Warning;
+    break;
+  case QUEUE_ERROR:
+    usedType = CGUIDialogKaiToast::Error;
+    break;
+  case QUEUE_INFO:
+  default:
+    usedType = CGUIDialogKaiToast::Info;
+    break;
+  }
+  CGUIDialogKaiToast::QueueNotification(usedType, caption, description, displayTime, withSound, messageTime);
+}
+
+void Interface_General::queue_notification_with_image(void* kodiBase, const char* imageFile, const char* caption,
+                                                      const char* description, unsigned int displayTime,
+                                                      bool withSound, unsigned int messageTime)
```

Comment:
I still don't get why we have three different queue_notification functions. All three function signatures are very similar. 

IMO, one function is sufficient:

<pre>
bool Interface_General::queue_notification(void* kodiBase, int type, const char* imageFile, const char* caption,
                                                      const char* description, unsigned int displayTime,
                                                      bool withSound, unsigned int messageTime)
</pre>

This covers all cases.

The boolean return value is useful for covering error cases (invalid paramter combinations, values, ...).